src/main/java/com/mparticle/kits/SingularKit.java

```diff
@@ -0,0 +1,252 @@
+
+    }
+
+    @Override
+    public boolean onPushRegistration(String deviceToken, String senderId) {
+        try {
+            switch (MPUtility.getAvailableInstanceId()) {
+                case GCM:
+                    Singular.setGCMDeviceToken(deviceToken);
+                    break;
+                case FCM:
+                    Singular.setFCMDeviceToken(deviceToken);
+                    break;
+            }
+        } catch (Exception unableToSetDeviceToken) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public void setInstallReferrer(Intent intent) {
+        new SingularInstallReceiver().onReceive(getContext(), intent);
+    }
+
+
+    @Override
+    public List<ReportingMessage> logLtvIncrease(BigDecimal bigDecimal, BigDecimal bigDecimal1, String s, Map<String, String> map) {
+        return null;
+    }
+
+    @Override
+    public List<ReportingMessage> logEvent(CommerceEvent commerceEvent) {
+        List<ReportingMessage> messages = new LinkedList<ReportingMessage>();
+        if (commerceEvent.getProductAction().equals(Product.PURCHASE)) {
+            if (!KitUtils.isEmpty(commerceEvent.getCurrency())) {
+                currency = commerceEvent.getCurrency();
+            }
+            if (commerceEvent.getProducts().size() > 0) {
+                List<Product> productList = commerceEvent.getProducts();
+                for (Product product : productList) {
+                    productPrice = product.getUnitPrice();
+                    productQuantity = product.getQuantity();
+                    if (!KitUtils.isEmpty(product.getSku())) {
+                        productSKU = product.getSku();
+                    }
+                    if (!KitUtils.isEmpty(product.getCategory())) {
+                        productCategory = product.getCategory();
+                    }
+                    productName = product.getName();
+                    amount = product.getTotalAmount();
+                    Singular.revenue(currency, amount, productSKU, productName, productCategory, (int) productQuantity, productPrice);
+                }
+                messages.add(ReportingMessage.fromEvent(this, commerceEvent));
+                return messages;
+            }
+        }
+        List<MPEvent> eventList = CommerceEventUtils.expand(commerceEvent);
+        if (eventList != null) {
+            for (int i = 0; i < eventList.size(); i++) {
+                try {
+                    logEvent(eventList.get(i));
+                    messages.add(ReportingMessage.fromEvent(this, commerceEvent));
+                } catch (Exception e) {
+                    Logger.warning("Failed to call logCustomEvent to Singular kit: " + e.toString());
+                }
+            }
+        }
+        return messages;
+    }
+
+    @Override
+    public void checkForDeepLink() {
+        config.withDDLTimeoutInSec(DDL_HANDLER_TIMEOUT_SEC);
```

Comment:
This still won't work as expected, unless calling `config.withDDLHandler` will trigger the `handleLink` callback to be invoked *again*. If that is not the case - then per my suggestion above, what I would suggest is just calling `handleLink` here, passing in the `mLink` value that was originally saved to memory on the original `handleLink` call.

**However** - I would also suggest setting the `link`value to null after passing it to `handleLink` here. The case to avoid, which is common when customers have multiple link-handling SDKs, is when the application is sent to the background and then a new link from another provider is tapped. Implementing developers will always test this scenario - it's crucial to get right. For example:
```
@Override
public void checkForDeepLink() {
    this.handleLink(mLink);
    mLink = null;
}
```